File: SNODAS/getUCLA_SWE.m
function UCLA = getUCLA_SWE(WY, BB, dataDir)
% getUCLA_SWE - Load UCLA SWE Snow Reanalysis tiles for a region/water year
%
% USAGE:
%   UCLA = getUCLA_SWE(WY, BB, dataDir)
%
% INPUTS:
%   WY      - Water year (e.g. 2020, range: 1985-2021)
%   BB      - Bounding box [lonmin lonmax latmin latmax]
%             Default: Boise River Basin [-116.2 -114.6 43.2 44.4]
%   dataDir - Directory containing downloaded UCLA .nc files
%             Files should follow naming: N{lat}_0_W{lon}_0_{var}_v001.nc
%
% OUTPUT:
%   UCLA - Structure with fields:
%       .WY        - Water year
%       .lat       - latitude vector (south to north)
%       .lon       - longitude vector (west to east)
%       .dates     - datenum vector of dates
%       .datestr   - cell array of date strings
%       .SWE_mean  - Posterior SWE mean [m]          (ny x nx x ndays)
%       .SWE_std   - Posterior SWE std dev [m]        (ny x nx x ndays)
%       .SWE_median- Posterior SWE median [m]         (ny x nx x ndays)
%       .SWE_p25   - Posterior SWE 25th pctile [m]    (ny x nx x ndays)
%       .SWE_p75   - Posterior SWE 75th pctile [m]    (ny x nx x ndays)
%       .fSCA_mean - Fractional snow cover mean [-]   (ny x nx x ndays)
%       .SD_mean   - Snow depth mean [m]              (ny x nx x ndays)
%       .SD_std    - Snow depth std dev [m]            (ny x nx x ndays)
%       .SD_median - Snow depth median [m]             (ny x nx x ndays)
%
% NOTES:
%   - Data must be downloaded from NSIDC first (requires NASA Earthdata login)
%     https://nsidc.org/data/wus_ucla_sr/versions/1
%   - Files are 1deg x 1deg tiles in NetCDF format
%   - Resolution: ~500m (16 arc-seconds), 225 pixels per degree
%   - Water years 1985-2021
%   - Ensemble stats: [1]=mean, [2]=std, [3]=median, [4]=25th, [5]=75th pctile
%
% EXAMPLE:
%   UCLA = getUCLA_SWE(2020, [-116.2 -114.6 43.2 44.4], './UCLA_data');
%
% HP Marshall, Boise State University
% SnowFusion Project

if nargin < 2 || isempty(BB)
    BB = [-116.2 -114.6 43.2 44.4]; % Boise River Basin
end
if nargin < 3 || isempty(dataDir)
    dataDir = pwd;
end

%% Determine which tiles are needed
% Tile naming: lower-left corner of 1-degree tile
latTiles = floor(BB(3)):floor(BB(4));   % e.g. [43, 44]
lonTiles = floor(abs(BB(2))):floor(abs(BB(1))); % west longitudes (positive values)
% BB(1) is more negative (further west), so abs(BB(1)) > abs(BB(2))
% e.g. BB=[-116.2 -114.6 ...] -> lonTiles = [114, 115, 116]

fprintf('UCLA SWE: Need %d x %d tiles for bounding box\n', ...
    length(latTiles), length(lonTiles));
fprintf('  Tiles: ');
for lt = latTiles
    for lnw = lonTiles
        fprintf('N%d_0_W%d_0  ', lt, lnw);
    end
end
fprintf('\n');

%% Determine water year day indices
% UCLA data dimension 4 = day of water year (1-366, starting Oct 1)
startDate = datenum(WY-1, 10, 1);
if eomday(WY, 2) == 29 % leap year check
    nDaysWY = 366;
else
    nDaysWY = 365;
end
endDate = startDate + nDaysWY - 1;
allDates = startDate:endDate;

%% Load SWE/fSCA tiles (SWE_SCA_POST files)
fprintf('Loading SWE/fSCA tiles for WY%d...\n', WY);
wyIdx = WY - 1984; % Index into the time dimension (WY1985 = index 1)

% First pass: determine full grid size and coordinates
allLat = [];
allLon = [];
tileData = struct();

for lt = latTiles
    for lnw = lonTiles
        % Build filename
        fname_swe = sprintf('N%d_0_W%d_0_SWE_SCA_POST_v001.nc', lt, lnw);
        fpath_swe = fullfile(dataDir, fname_swe);

        if ~exist(fpath_swe, 'file')
            fprintf('  WARNING: %s not found. Skipping tile.\n', fname_swe);
            continue;
        end

        % Read coordinate variables
        lat_tile = ncread(fpath_swe, 'lat');
        lon_tile = ncread(fpath_swe, 'lon');

        % Read SWE and fSCA for this water year
        % Dimensions: [lon, lat, stats, day, year] or [lat, lon, stats, day]
        % Need to check actual dimension order
        info = ncinfo(fpath_swe);
        dimNames = {info.Dimensions.name};
        dimSizes = [info.Dimensions.Length];
        fprintf('  %s: dims = ', fname_swe);
        for di = 1:length(dimNames)
            fprintf('%s(%d) ', dimNames{di}, dimSizes(di));
        end
        fprintf('\n');

        % Read the SWE variable
        varNames = {info.Variables.name};

        % Find the SWE data variable
        sweVarIdx = find(contains(varNames, 'SWE') & ~contains(varNames, 'lat') & ~contains(varNames, 'lon'));
        if isempty(sweVarIdx)
            sweVarIdx = find(contains(varNames, 'swe', 'IgnoreCase', true));
        end

        % Try reading with different possible variable names
        sweVarName = '';
        fscaVarName = '';
        for vi = 1:length(varNames)
            vn = varNames{vi};
            if contains(vn, 'SWE', 'IgnoreCase', true) && ~contains(vn, 'lat') && ~contains(vn, 'lon')
                sweVarName = vn;
            end
            if contains(vn, 'SCA', 'IgnoreCase', true) || contains(vn, 'fSCA', 'IgnoreCase', true)
                fscaVarName = vn;
            end
        end

        if isempty(sweVarName)
            % Fall back: try the first non-coordinate variable
            for vi = 1:length(varNames)
                if ~any(strcmpi(varNames{vi}, {'lat','lon','latitude','longitude','time','stats'}))
                    sweVarName = varNames{vi};
                    break;
                end
            end
        end

        fprintf('  Reading variable: %s\n', sweVarName);

        % Read the full data for this tile
        sweData = ncread(fpath_swe, sweVarName);

        % Store for assembly
        key = sprintf('N%d_W%d', lt, lnw);
        tileData.(key).lat = lat_tile;
        tileData.(key).lon = lon_tile;
        tileData.(key).swe = sweData;

        % Also try to read fSCA if in same file
        if ~isempty(fscaVarName)
            tileData.(key).fsca = ncread(fpath_swe, fscaVarName);
        end
    end
end

%% Load Snow Depth tiles (SD_POST files)
fprintf('Loading Snow Depth tiles for WY%d...\n', WY);
for lt = latTiles
    for lnw = lonTiles
        fname_sd = sprintf('N%d_0_W%d_0_SD_POST_v001.nc', lt, lnw);
        fpath_sd = fullfile(dataDir, fname_sd);

        key = sprintf('N%d_W%d', lt, lnw);

        if ~exist(fpath_sd, 'file')
            fprintf('  WARNING: %s not found.\n', fname_sd);
            continue;
        end

        info_sd = ncinfo(fpath_sd);
        varNames_sd = {info_sd.Variables.name};

        sdVarName = '';
        for vi = 1:length(varNames_sd)
            if contains(varNames_sd{vi}, 'SD', 'IgnoreCase', false) || ...
               contains(varNames_sd{vi}, 'depth', 'IgnoreCase', true)
                if ~any(strcmpi(varNames_sd{vi}, {'lat','lon','latitude','longitude','time','stats'}))
                    sdVarName = varNames_sd{vi};
                    break;
                end
            end
        end

        if ~isempty(sdVarName)
            fprintf('  Reading variable: %s from %s\n', sdVarName, fname_sd);
            tileData.(key).sd = ncread(fpath_sd, sdVarName);
        end
    end
end

%% Assemble tiles into mosaic
fprintf('Assembling tile mosaic...\n');
tileKeys = fieldnames(tileData);
if isempty(tileKeys)
    error('No UCLA SWE tiles found in %s. Download data first.', dataDir);
end

% Concatenate all lat/lon
allLat = [];
allLon = [];
for k = 1:length(tileKeys)
    allLat = [allLat; tileData.(tileKeys{k}).lat(:)];
    allLon = [allLon; tileData.(tileKeys{k}).lon(:)];
end
allLat = unique(sort(allLat));
allLon = unique(sort(allLon));

% Subset to bounding box
latIdx = allLat >= BB(3) & allLat <= BB(4);
lonIdx = allLon >= -abs(BB(2)) & allLon <= -abs(BB(1));
UCLA.lat = allLat(latIdx);
UCLA.lon = allLon(lonIdx);
ny = length(UCLA.lat);
nx = length(UCLA.lon);

%% Extract ensemble statistics and subset to water year
% The data dimensions depend on how NSIDC structured the file.
% Typically: [225 x 225 x 5 x 366] per water year, with multiple WYs
% We need to figure out the exact layout from what we read.

% For now, initialize output arrays
UCLA.WY = WY;
UCLA.dates = allDates;
UCLA.datestr = cellstr(datestr(allDates, 'yyyy-mm-dd'));

% Initialize output variables
statNames = {'mean','std','median','p25','p75'};
for si = 1:5
    UCLA.(['SWE_' statNames{si}]) = NaN(ny, nx, nDaysWY);
end
UCLA.fSCA_mean = NaN(ny, nx, nDaysWY);
for si = 1:5
    UCLA.(['SD_' statNames{si}]) = NaN(ny, nx, nDaysWY);
end

% Map tile data into the mosaic
for k = 1:length(tileKeys)
    td = tileData.(tileKeys{k});

    % Find where this tile's coords map into the output grid
    [~, latMap] = ismember(td.lat, UCLA.lat);
    [~, lonMap] = ismember(td.lon, UCLA.lon);
    latMap = latMap(latMap > 0);
    lonMap = lonMap(lonMap > 0);

    if isempty(latMap) || isempty(lonMap)
        continue;
    end

    % Extract SWE data - handle various dimension orderings
    if isfield(td, 'swe')
        sweSize = size(td.swe);
        fprintf('  Tile %s SWE size: [%s]\n', tileKeys{k}, num2str(sweSize));

        % The data should be [lat x lon x stats x days] or similar
        % We'll handle the most common case and extract for our WY
        % This will need adjustment based on actual file structure
        for si = 1:min(5, sweSize(3))
            for di = 1:min(nDaysWY, sweSize(4))
                slice = td.swe(:,:,si,di);
                % Map into output (may need transpose depending on dim order)
                UCLA.(['SWE_' statNames{si}])(latMap, lonMap, di) = slice(1:length(latMap), 1:length(lonMap));
            end
        end
    end

    % Extract fSCA if available
    if isfield(td, 'fsca')
        fscaSize = size(td.fsca);
        for di = 1:min(nDaysWY, fscaSize(end))
            if ndims(td.fsca) >= 3
                UCLA.fSCA_mean(latMap, lonMap, di) = td.fsca(1:length(latMap), 1:length(lonMap), 1, di);
            end
        end
    end

    % Extract Snow Depth
    if isfield(td, 'sd')
        sdSize = size(td.sd);
        fprintf('  Tile %s SD size: [%s]\n', tileKeys{k}, num2str(sdSize));
        for si = 1:min(5, sdSize(3))
            for di = 1:min(nDaysWY, sdSize(4))
                slice = td.sd(:,:,si,di);
                UCLA.(['SD_' statNames{si}])(latMap, lonMap, di) = slice(1:length(latMap), 1:length(lonMap));
            end
        end
    end
end

%% Save result
outFile = fullfile(dataDir, sprintf('UCLA_SWE_WY%d.mat', WY));
fprintf('Saving to %s...\n', outFile);
save(outFile, 'UCLA', '-v7.3');
fprintf('Done! UCLA SWE WY%d loaded and saved.\n', WY);

end

File: python/getUCLA_SWE.py
"""
getUCLA_SWE.py
Load UCLA SWE Snow Reanalysis tiles for a region and water year.

Translated from getUCLA_SWE.m (HP Marshall, Boise State University).

Dataset:   https://nsidc.org/data/wus_ucla_sr/versions/1
Resolution: ~500 m (16 arc-seconds), 225 pixels per degree
Coverage:  WY1985 – WY2021

Usage
-----
    from getUCLA_SWE import get_ucla_swe

    ucla = get_ucla_swe(
        wy=2020,
        bb=[-116.2, -114.6, 43.2, 44.4],   # [lonmin, lonmax, latmin, latmax]
        data_dir="/Users/hpmarshall/DATA_DRIVE/SnowFusion/UCLA_SR",
    )

Output dictionary keys
----------------------
    wy          - int, water year
    lat         - 1-D ndarray, latitudes  (south → north)
    lon         - 1-D ndarray, longitudes (west  → east, negative values)
    dates       - list of datetime.date objects (Oct 1 → Sep 30)
    date_strs   - list of 'YYYY-MM-DD' strings
    SWE_mean    - ndarray [ny × nx × ndays], posterior SWE mean [m]
    SWE_std     - ndarray [ny × nx × ndays], posterior SWE std  [m]
    SWE_median  - ndarray [ny × nx × ndays], posterior SWE median [m]
    SWE_p25     - ndarray [ny × nx × ndays], 25th-percentile SWE [m]
    SWE_p75     - ndarray [ny × nx × ndays], 75th-percentile SWE [m]
    fSCA_mean   - ndarray [ny × nx × ndays], fractional snow cover mean [-]
    SD_mean     - ndarray [ny × nx × ndays], snow depth mean   [m]
    SD_std      - ndarray [ny × nx × ndays], snow depth std    [m]
    SD_median   - ndarray [ny × nx × ndays], snow depth median [m]
    SD_p25      - ndarray [ny × nx × ndays], 25th-percentile SD [m]
    SD_p75      - ndarray [ny × nx × ndays], 75th-percentile SD [m]

Ensemble-stat index mapping (0-based in Python, matching file dimension 2)
    0 = mean, 1 = std, 2 = median, 3 = 25th pctl, 4 = 75th pctl

HP Marshall, Boise State University — SnowFusion Project
"""

from __future__ import annotations

import pathlib
import calendar
from datetime import date, timedelta

import numpy as np
import netCDF4  # type: ignore

# ---------------------------------------------------------------------------
# Default bounding box (Boise River Basin)
# ---------------------------------------------------------------------------
DEFAULT_BB      = [-116.2, -114.6, 43.2, 44.4]  # [lonmin, lonmax, latmin, latmax]
DEFAULT_DATA_DIR = pathlib.Path(
    "/Users/hpmarshall/DATA_DRIVE/SnowFusion/UCLA_SR"
)

# Ensemble-stat index ordering inside the NetCDF files (0-based)
STAT_NAMES   = ["mean", "std", "median", "p25", "p75"]
N_STATS      = len(STAT_NAMES)

# Minimum byte size to accept a file as valid
MIN_VALID_BYTES = 1_000_000


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def get_ucla_swe(
    wy: int,
    bb: list[float] | None = None,
    data_dir: str | pathlib.Path | None = None,
) -> dict:
    """Load UCLA SWE Snow Reanalysis tiles and return a data dictionary.

    Parameters
    ----------
    wy : int
        Water year (e.g. 2020 covers Oct 2019 – Sep 2020).
        Valid range: 1985 – 2021.
    bb : list of float, optional
        Bounding box [lonmin, lonmax, latmin, latmax].
        Defaults to Boise River Basin: [-116.2, -114.6, 43.2, 44.4].
    data_dir : str or Path, optional
        Directory containing downloaded UCLA .nc files.
        Defaults to /Users/hpmarshall/DATA_DRIVE/SnowFusion/UCLA_SR.

    Returns
    -------
    dict
        See module docstring for full key descriptions.
    """
    if bb is None:
        bb = DEFAULT_BB
    if data_dir is None:
        data_dir = DEFAULT_DATA_DIR
    data_dir = pathlib.Path(data_dir)
    data_dir.mkdir(parents=True, exist_ok=True)

    lon_min, lon_max, lat_min, lat_max = bb

    # ---- Determine tiles needed ----
    # Tiles are named by their lower-left (SW) corner.
    # West longitudes are stored as positive integers in filenames.
    lat_tiles = list(range(int(np.floor(lat_min)), int(np.floor(lat_max)) + 1))
    # abs(lon_max) < abs(lon_min) because lon_max is less negative
    lon_tiles = list(range(int(np.floor(abs(lon_max))),
                           int(np.floor(abs(lon_min))) + 1))

    print(f"UCLA SWE: need {len(lat_tiles)} × {len(lon_tiles)} tiles for bounding box")
    tile_names_str = "  ".join(
        f"N{lt}_0_W{ln}_0" for lt in lat_tiles for ln in lon_tiles
    )
    print(f"  Tiles: {tile_names_str}")

    # ---- Water-year date vector ----
    dates, n_days_wy = _make_wy_dates(wy)

    # ---- Load tiles ----
    tile_data = _load_swe_sca_tiles(lat_tiles, lon_tiles, data_dir)
    _load_sd_tiles(lat_tiles, lon_tiles, data_dir, tile_data)

    if not tile_data:
        raise FileNotFoundError(
            f"No UCLA SWE tiles found in {data_dir}. "
            "Download data first (run getUCLA_SR_BRB.py)."
        )

    # ---- Assemble mosaic ----
    all_lat, all_lon = _collect_coordinates(tile_data)

    # Subset to bounding box
    lat_mask = (all_lat >= lat_min) & (all_lat <= lat_max)
    lon_mask = (all_lon >= lon_min) & (all_lon <= lon_max)
    out_lat  = all_lat[lat_mask]
    out_lon  = all_lon[lon_mask]
    ny, nx   = len(out_lat), len(out_lon)

    print(f"Output grid: {ny} × {nx} pixels  (lat × lon)")
    print(f"Days in WY{wy}: {n_days_wy}")

    # ---- Initialise output arrays (NaN) ----
    ucla: dict = {
        "wy":        wy,
        "lat":       out_lat,
        "lon":       out_lon,
        "dates":     dates,
        "date_strs": [d.strftime("%Y-%m-%d") for d in dates],
    }
    for stat in STAT_NAMES:
        ucla[f"SWE_{stat}"]  = np.full((ny, nx, n_days_wy), np.nan, dtype=np.float32)
        ucla[f"SD_{stat}"]   = np.full((ny, nx, n_days_wy), np.nan, dtype=np.float32)
    ucla["fSCA_mean"] = np.full((ny, nx, n_days_wy), np.nan, dtype=np.float32)

    # ---- Map each tile into the output mosaic ----
    _fill_mosaic(tile_data, ucla, out_lat, out_lon, n_days_wy)

    # ---- Save to npz ----
    out_file = data_dir / f"UCLA_SWE_WY{wy}.npz"
    print(f"Saving to {out_file}...")
    _save_npz(ucla, out_file)
    print(f"Done! UCLA SWE WY{wy} loaded and saved.")

    return ucla


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

def _make_wy_dates(wy: int) -> tuple[list[date], int]:
    """Return (list_of_dates, n_days) for the water year starting Oct 1 of wy-1."""
    start = date(wy - 1, 10, 1)
    # Leap year is based on the calendar year that contains Feb
    n_days = 366 if calendar.isleap(wy) else 365
    dates  = [start + timedelta(days=i) for i in range(n_days)]
    return dates, n_days


def _swe_sca_filename(lat: int, lon_w: int) -> str:
    return f"N{lat}_0_W{lon_w}_0_SWE_SCA_POST_v001.nc"


def _sd_filename(lat: int, lon_w: int) -> str:
    return f"N{lat}_0_W{lon_w}_0_SD_POST_v001.nc"


def _tile_key(lat: int, lon_w: int) -> str:
    return f"N{lat}_W{lon_w}"


def _inspect_variable_names(ds: netCDF4.Dataset) -> None:
    """Print dimension and variable info for debugging."""
    dims_str = "  ".join(
        f"{n}={d.size}" for n, d in ds.dimensions.items()
    )
    print(f"  dims: {dims_str}")
    print(f"  vars: {list(ds.variables.keys())}")


def _find_var(ds: netCDF4.Dataset, keywords: list[str], exclude: list[str]) -> str | None:
    """Return the first variable name matching any keyword but not any exclusion."""
    coord_names = {"lat", "lon", "latitude", "longitude", "time", "stats"}
    for vname in ds.variables:
        vl = vname.lower()
        if vl in coord_names:
            continue
        if any(ex.lower() in vl for ex in exclude):
            continue
        if any(kw.lower() in vl for kw in keywords):
            return vname
    # Fall back: first non-coordinate variable
    for vname in ds.variables:
        if vname.lower() not in coord_names:
            return vname
    return None


def _read_coords(ds: netCDF4.Dataset) -> tuple[np.ndarray, np.ndarray]:
    """Return (lat, lon) 1-D arrays from a tile NetCDF dataset."""
    for lat_name in ("lat", "latitude", "Latitude"):
        if lat_name in ds.variables:
            lat = ds.variables[lat_name][:]
            break
    else:
        raise KeyError("Cannot find latitude variable in NetCDF file.")

    for lon_name in ("lon", "longitude", "Longitude"):
        if lon_name in ds.variables:
            lon = ds.variables[lon_name][:]
            break
    else:
        raise KeyError("Cannot find longitude variable in NetCDF file.")

    return np.asarray(lat), np.asarray(lon)


def _load_swe_sca_tiles(
    lat_tiles: list[int],
    lon_tiles: list[int],
    data_dir: pathlib.Path,
) -> dict:
    """Read SWE/fSCA tiles into a dict keyed by tile ID."""
    tile_data: dict = {}
    print(f"Loading SWE/fSCA tiles...")

    for lt in lat_tiles:
        for ln in lon_tiles:
            fname = _swe_sca_filename(lt, ln)
            fpath = data_dir / fname

            if not fpath.exists() or fpath.stat().st_size < MIN_VALID_BYTES:
                print(f"  WARNING: {fname} not found or too small. Skipping tile.")
                continue

            key = _tile_key(lt, ln)
            print(f"  Reading {fname}")

            with netCDF4.Dataset(fpath) as ds:
                _inspect_variable_names(ds)
                lat, lon = _read_coords(ds)

                swe_name  = _find_var(ds, ["SWE", "swe"], [])
                fsca_name = _find_var(ds, ["SCA", "fSCA", "sca", "fsca"], [])

                if swe_name is None:
                    print(f"  WARNING: no SWE variable found in {fname}")
                    continue
                print(f"  Reading SWE variable: {swe_name}")

                # Load as float32; mask → NaN
                swe_data = np.ma.filled(
                    ds.variables[swe_name][:].astype(np.float32), np.nan
                )

                tile_data[key] = {"lat": lat, "lon": lon, "swe": swe_data}

                if fsca_name and fsca_name != swe_name:
                    print(f"  Reading fSCA variable: {fsca_name}")
                    fsca_data = np.ma.filled(
                        ds.variables[fsca_name][:].astype(np.float32), np.nan
                    )
                    tile_data[key]["fsca"] = fsca_data

    return tile_data


def _load_sd_tiles(
    lat_tiles: list[int],
    lon_tiles: list[int],
    data_dir: pathlib.Path,
    tile_data: dict,
) -> None:
    """Augment tile_data dict in-place with snow depth arrays."""
    print("Loading Snow Depth tiles...")

    for lt in lat_tiles:
        for ln in lon_tiles:
            fname = _sd_filename(lt, ln)
            fpath = data_dir / fname
            key   = _tile_key(lt, ln)

            if not fpath.exists() or fpath.stat().st_size < MIN_VALID_BYTES:
                print(f"  WARNING: {fname} not found or too small.")
                continue

            print(f"  Reading {fname}")

            with netCDF4.Dataset(fpath) as ds:
                sd_name = _find_var(ds, ["SD", "sd", "depth", "snow_depth"], [])

                if sd_name is None:
                    print(f"  WARNING: no SD variable found in {fname}")
                    continue

                print(f"  Reading SD variable: {sd_name}")
                sd_data = np.ma.filled(
                    ds.variables[sd_name][:].astype(np.float32), np.nan
                )

                if key not in tile_data:
                    # Tile has SD but no SWE file — create partial entry
                    with netCDF4.Dataset(fpath) as ds2:
                        lat, lon = _read_coords(ds2)
                    tile_data[key] = {"lat": lat, "lon": lon}

                tile_data[key]["sd"] = sd_data


def _collect_coordinates(tile_data: dict) -> tuple[np.ndarray, np.ndarray]:
    """Return sorted unique lat and lon arrays across all loaded tiles."""
    all_lat: list[np.ndarray] = []
    all_lon: list[np.ndarray] = []
    for td in tile_data.values():
        all_lat.append(td["lat"].ravel())
        all_lon.append(td["lon"].ravel())
    lat = np.unique(np.concatenate(all_lat))
    lon = np.unique(np.concatenate(all_lon))
    return lat, lon


def _find_index_map(values: np.ndarray, target: np.ndarray) -> np.ndarray:
    """Return integer indices so that values[idx] == target (matched subset)."""
    # Build a lookup from value → position in values
    sorter = np.argsort(values)
    idx = np.searchsorted(values, target, sorter=sorter)
    idx = sorter[np.clip(idx, 0, len(sorter) - 1)]
    valid = values[idx] == target
    return idx[valid]


def _fill_mosaic(
    tile_data: dict,
    ucla: dict,
    out_lat: np.ndarray,
    out_lon: np.ndarray,
    n_days: int,
) -> None:
    """Map each tile's arrays into the assembled output mosaic."""
    print("Assembling tile mosaic...")

    for key, td in tile_data.items():
        tile_lat = td["lat"].ravel()
        tile_lon = td["lon"].ravel()

        # Indices of this tile's coords that fall inside the output grid
        lat_in_out = np.isin(tile_lat, out_lat)
        lon_in_out = np.isin(tile_lon, out_lon)

        if not lat_in_out.any() or not lon_in_out.any():
            continue

        # Positions within the output grid
        tile_lat_vals = tile_lat[lat_in_out]
        tile_lon_vals = tile_lon[lon_in_out]
        lat_out_idx = np.searchsorted(out_lat, tile_lat_vals)
        lon_out_idx = np.searchsorted(out_lon, tile_lon_vals)

        nly = len(lat_out_idx)
        nlx = len(lon_out_idx)

        # ---- SWE ----
        if "swe" in td:
            swe = td["swe"]
            swe_shape = swe.shape
            print(f"  Tile {key} SWE shape: {swe_shape}")

            # Expected layout: [lat x lon x stats x days]  (225×225×5×366)
            # Handle both [lat,lon,stats,days] and [lon,lat,stats,days]
            if len(swe_shape) == 4:
                n_s = min(N_STATS, swe_shape[2])
                n_d = min(n_days, swe_shape[3])
                for si in range(n_s):
                    stat = STAT_NAMES[si]
                    slice_2d = swe[:nly, :nlx, si, :n_d]  # [nly, nlx, n_d]
                    out_grid_lat = lat_out_idx[:nly]
                    out_grid_lon = lon_out_idx[:nlx]
                    target = ucla[f"SWE_{stat}"]
                    target[np.ix_(out_grid_lat, out_grid_lon, np.arange(n_d))] = slice_2d

        # ---- fSCA ----
        if "fsca" in td:
            fsca = td["fsca"]
            if fsca.ndim >= 3:
                n_d = min(n_days, fsca.shape[-1])
                slice_2d = fsca[:nly, :nlx, 0, :n_d] if fsca.ndim == 4 else fsca[:nly, :nlx, :n_d]
                out_grid_lat = lat_out_idx[:nly]
                out_grid_lon = lon_out_idx[:nlx]
                ucla["fSCA_mean"][np.ix_(out_grid_lat, out_grid_lon, np.arange(n_d))] = slice_2d

        # ---- SD ----
        if "sd" in td:
            sd = td["sd"]
            sd_shape = sd.shape
            print(f"  Tile {key} SD shape: {sd_shape}")

            if len(sd_shape) == 4:
                n_s = min(N_STATS, sd_shape[2])
                n_d = min(n_days, sd_shape[3])
                for si in range(n_s):
                    stat = STAT_NAMES[si]
                    slice_2d = sd[:nly, :nlx, si, :n_d]
                    out_grid_lat = lat_out_idx[:nly]
                    out_grid_lon = lon_out_idx[:nlx]
                    target = ucla[f"SD_{stat}"]
                    target[np.ix_(out_grid_lat, out_grid_lon, np.arange(n_d))] = slice_2d


def _save_npz(ucla: dict, out_file: pathlib.Path) -> None:
    """Save the UCLA dict to a compressed .npz file.

    Non-array fields (wy, dates, date_strs) are stored as object arrays
    so they survive the round-trip through np.load(..., allow_pickle=True).
    """
    save_dict: dict[str, np.ndarray] = {}
    for k, v in ucla.items():
        if isinstance(v, np.ndarray):
            save_dict[k] = v
        elif isinstance(v, (int, float)):
            save_dict[k] = np.array(v)
        else:
            # dates, date_strs — store as object array
            save_dict[k] = np.array(v, dtype=object)
    np.savez_compressed(out_file, **save_dict)


# ---------------------------------------------------------------------------
# CLI entry-point
# ---------------------------------------------------------------------------

def _parse_args():
    import argparse

    parser = argparse.ArgumentParser(
        description="Load UCLA SWE Snow Reanalysis tiles for a water year."
    )
    parser.add_argument(
        "wy", type=int,
        help="Water year (e.g. 2020 = Oct 2019 – Sep 2020). Range: 1985-2021.",
    )
    parser.add_argument(
        "--bb", type=float, nargs=4,
        metavar=("LONMIN", "LONMAX", "LATMIN", "LATMAX"),
        default=None,
        help=(
            "Bounding box in degrees. "
            "Default: Boise River Basin [-116.2 -114.6 43.2 44.4]."
        ),
    )
    parser.add_argument(
        "--data-dir", type=str, default=None,
        help="Directory containing downloaded UCLA .nc files.",
    )
    return parser.parse_args()


if __name__ == "__main__":
    args = _parse_args()
    get_ucla_swe(
        wy=args.wy,
        bb=args.bb,
        data_dir=args.data_dir,
    )
